### ISSUE TITLE
Suggests use of @property

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#200]
_It's a bit complicated, but let's go..._

When you are using the `__get()` magic method, is hard you get the typehint of your magic property, because IDE can't knows that before the runtime (and IDE can't do that). But, we can tell before, _because we know_, by using the  `@property` annotation on class.

Things like `$user->created_at->lt(...)` turns harder, because `created_at` are not a real property, then IDE can't know that it is a `Carbon` instance, then don't helps with `lt()` methods and others. 

**To solve it:**

```php
/**
 * @property Carbon $created_at
 */
class User extends Model {
}
```

The major problem here is that we should do that for each property that we have. Of course, the correct way is we defines that while we are writing the class, but in some development methods, it is done only on end (_or never is done_).

To solve this problem, I suggests the implementation of an inspection that tells that you should declares the `@property` on class and help it with an autofix. But let's do that by steps:

---

**1. Suggests defines the `@property` annotation:**

To make it work, EA will detects that the property is not defined on instance class, but it have the `__get()` method on reading or `__set()` method on writing.

```php
$user = new User;
echo $user->created_at->format('Y-m-d');
// Warning: created_at property is not defined on User class.
// Solution: declare it or annotate with @property.
```

---

**2. Suggests an autofixer on writing:**

When we are writing on the non-declared property, we could know exactly what it is expected to be. Like is done with `@param`s. Then:

```php
$user = new User;
$user->created_at = Carbon::now();
// Warning: created_at property is not defined on User class.
// Autofix: annotate with @property Carbon.
```

---

**3. Additionally, it should detects that this property is multitype:**

If your property could be represented by a lot of differents types, it should be merge existent types with new types. But it should avoid that by warning user and will give two autofixers:

```php
/** [User] @property Carbon $created_at */
$user = new User;
$user->created_at = 'hello';
// Warning: created_at property is defined as Carbon, string given.
// Autofix: annotate with @property Carbon|string.
// Autofix: annotate with @property string only.
```

---

**4. Additionally, it should prefer top class instead of parent class:**

In some cases, maybe you are working with two classes where the second is a child class. For instance, `DateTime` is the parent class of [`Carbon`](https://github.com/briannesbitt/Carbon/blob/master/src/Carbon/Carbon.php#L53). In this case, it should suggests that you replaces current by a child class.

```php
/** [User] @property DateTime $created_at */
$user = new User;
$user->created_at = DateTime::createFromFormat(...); // Matches with @property.
$user->created_at = Carbon::now();
// Warning: created_at property is defined as DateTime, Carbon given.
// Autofix: upgrade annotation with @property Carbon. (will drop DateTime)
// Autofix: annotate with @property DateTime|Carbon. (will keep both)
```

---

**5. Detects property type on reading.**

It should be the harder case. Basically, this inspection should detects how the property is treated on reading and suggests how it should be annotated. I will try explain in on practice.

```php
$user = new User;
echo $user->created_at->format('Y-m-d');
```

Is harder you understand, without annotation, that this `created_at` property is a Carbon instance, because you can have the `format()` method in a lot of other classes (like in `DateInterval`, `DateTime`, etc). In this case, the unique thing that we can do is check if possible methods are defined like `format(string[, ...])`, then suggests about three classes that matches that. But still harder.

But, we can make a double check over this property by seeking another calls like that, and make a more refined filter. For instance:

```php
$user = new User;
echo $user->created_at->format('Y-m-d');
echo $user->created_at->gte(Carbon::now());
```

Which other class have `format()` and `gte()` methods together? In this case, is clear that we are talking about `Carbon` instance, then now we can suggests it to be annotated.

I suggests that only suggests that if we have until 5 possible classes after filter.

```php
$user = new User;
echo $user->created_at->format('Y-m-d');
echo $user->created_at->gte(Carbon::now());
// Warning: created_at property is not declared on User class.
// Autofix: annotate with @property Carbon.
```

Note: it should works with properties too. If you have a declared property on a class and you try read that, then it should detects what is your class.

```php
$user = new User;
echo $user->date_interval->y . 
     $user->date_interval->m . 
     $user->date_interval->d;
// Warning: date_interval property is not declared on User class.
// Autofix: annotate with @property DateInterval.
```

---

**6. Additionally, suggests the most parent class:**

In case where you can have some parents and child classes, suggests the most parent one that is compatible with the filter (it will make your property more flexible). For instance: both `Carbon` and `DateTime` have the methods `add()` and `sub()`. In this case, `DateTime` should be suggested, expect if you too use the `startOfDay()` that is a `Carbon` only method. In this same line, abstract classes or traits will have more value than child classes.

```php
$user = new User;
$user->created_at->add(...);
$user->created_at->sub(...);
// Warning: created_at property is not declared on User class.
// Autofix: annotate with @property DateTime. (instead of compatible Carbon)
```

---

Finally, I guess that the hardest case is the filter one, because you should try analyzes each methods and properties declared on possible classes, one by one, to try identify if the property could matches with your declaration.

**Comment by funivan** [comment]
In our project we try to remove usage of virtual `@property`. This is not a good approach to write dynamic code.

**Comment by rentalhost** [comment]
@funivan how do you treat public or magic properties?

**Comment by funivan** [comment]
@rentalhost we have old code with `@property` but we try to avoid that. In new classes we write private properties, getter and setter for them. It is easy to write and maintain classes with defined properties alongside with magic properties)

**Comment by rentalhost** [comment]
@funivan Ah, it's basically what Java devs does. I just don't know if in PHP it's fully recommended, once that it will decrease the code performance (_functions call is about 500% heavier_). In java it makes sense because it is compiled, then probabily will not cost more.

In all cases, I think that magic properties are very common mainly if you are using some framework. Then except if you code like your show, it is a valid case (in this case, I suspect that you don't declares the `__get()` method in general, which invalidates this inspection).

**Comment by funivan** [comment]
@rentalhost yes but `__get` is heavy method too. Also if we have `parent::__get()` calls ;)

**Comment by kalessil** [comment]
Oh that's PhpDoc, I have to skip this.
